pb/rpc: add google.rpc.ErrorInfo into rpc exceptions

rockwotj opened the pull request:
Support using structured errors to return information back to clients.

## Backports Required

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

rockwotj committed:
pb/rpc: support error_info in rpc errors

This way we will be able to pass back structured data in our errors.

See: https://connectrpc.com/docs/protocol#error-end-stream
See: https://github.com/googleapis/api-common-protos/blob/3332dec527759859840a3a2ff108c67a54708130/google/rpc/error_details.proto#L110

rockwotj committed:
admin/server: use simplified body writer in seastar http reply

Newer seastar versions has this new body writer that keeps the output
stream alive, so we no longer need to use `do_with` to keep the stream
alive or close it.

rockwotj committed:
admin/proxy: support passing back error_info

We should preserve error info when proxying rpc failures

rockwotj committed:
serde/pb/rpc: remove some forward declarations

These are not used

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--95bbbe5d-f438-4f2d-95b0-4f728b71b0e0-->
<!--TEST RESULTS-->
<!--2950d739-433b-4cfc-9a35-4c474e3ec191-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71412">build#71412</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71412#0198e79e-0f3b-4f6a-95b0-1efb849a81a9 | FLAKY         | 18/21    | upstream reliability is '91.71461449942463'. current run reliability is '85.71428571428571'. drift is 6.00033 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71430">build#71430</a></summary>


| test_class               | test_method                          | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-------------------------|:-------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest    | test_creating_and_listing_migrations | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/71430#0198e974-3f74-45b9-a83d-ac066ea02f69 | FLAKY         | 19/21    | upstream reliability is '99.2'. current run reliability is '90.47619047619048'. drift is 8.72381 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest | test_node_operations                 | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71430#0198e972-109b-4b59-8045-fd6f98429398 | FLAKY         | 19/21    | upstream reliability is '99.64912280701755'. current run reliability is '90.47619047619048'. drift is 9.17293 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

rockwotj committed:
serde/json: fix base64 encoding in writer

We were calling the decoding function, we want to call the encoding
function instead.

rockwotj committed:
pb/rpc: add test for error details

Check both propagating from another node and returning the error
directly works.

rockwotj commented:
FYI @r-vasquez I don't know if we will have any ErrorInfo usages for DR, but we'll have this ability to pass back structured errors.

pgellert commented on src/v/redpanda/admin/proxy/types.h:
```diff
@@ -75,13 +79,32 @@ struct proxy_response
     iobuf payload;
 
-    auto serde_fields() { return std::tie(error_code, payload); }
+    struct error_info
```
What's the reasoning behind why we need a redefinition of `serde::pb::rpc::error_info` here? Could we just make `serde::pb::rpc::error_info` serde serializable instead?

rockwotj replied:
We could, that rpc header is included for all services however. This file is only included in the .cc files of the proxy classes. I also think it's confusing to have serde stuff mixed in with protobuf code, so it feels cleaner to just isolated this.

rockwotj replied:
I will clarify in the wiki the expected usage

pgellert commented on src/v/serde/protobuf/rpc.h:
```diff
@@ -114,14 +115,74 @@ class base_service {
+    // source of the error. Example: "pubsub.googleapis.com". If the error is
+    // common across many APIs, the first segment of the example above will be
+    // omitted.  The value will be, "googleapis.com".
+    ss::sstring domain = ss::sstring(redpanda_core_domain);
```
Is the idea that we'll leave this `domain` field at its default or would each RPC throw within its own `redpanda_core_domain + "/my_rpc"` domain?

rockwotj replied:
This is a good question. No I expect us to not touch domain, but instead just carve out a list of reasons per subsystem with a few global ones. We want to expose a single view external to the admin server, and domain is really to separate between an error from different services, not subsystems.

rockwotj replied:
I will clarify expected usage in the wiki

pgellert reviewed (approved):
Looks great!

I assume the [EndStreamResponse](https://connectrpc.com/docs/protocol/#error-end-stream) does not apply to us because I suspect we won't be using streaming rpc's but only regular "unary" rpc's. Am I right there?

rockwotj commented:
> Looks great!
> 
> 
> 
> I assume the [EndStreamResponse](https://connectrpc.com/docs/protocol/#error-end-stream) does not apply to us because I suspect we won't be using streaming rpc's but only regular "unary" rpc's. Am I right there?

Correct. I don't see a strict need for streaming RPCs right now. They are more complex so I have been omitting them for now
